### ISSUE TITLE
Discussion on Visual feature aggregation

### ISSUE THREAD

**Issue by vidit98** [microsoft/TRELLIS#93]
Thanks again for releasing the code. I was going through [extract_features.py](https://github.com/microsoft/TRELLIS/blob/68820295a6ff17b44117d7439d0a244bd9c7826e/dataset_toolkits/extract_feature.py#L139) and found that to calculate the feature for a voxel, visibility of the voxel is not considered from a given camera.

Intuitively, I thought that features would be averaged only from rendered views from which a particular voxel is available. Do have any explanation why does visibility does not matter? Did you guys tried something like this before finalizing the current method?

**Comment by YuDeng** [comment]
This is a good question. We haven't tried taking visibility into consideration. 

In our case, a simple average over all views works pretty well. We conjecture that when averaging, the visible features corresponding to a voxel tend to dominate, given that objects are usually somewhat convex. Of course, we believe that incorporating visibility could yield even better results, and it is worth trying.

**Comment by cyw-3d** [comment]
I tried incorporating visibility into DINO feature aggregation using the PyTorch3D point cloud rasterizer to calculate visibility. I tested VAE reconstruction on 15 views, with visibility-based aggregation shown on the right of the figure and average aggregation on the left. The results appear quite similar. I suspect this is because the pretrained VAE I used was trained on average aggregation data, which contains multiple sources of noise. If a VAE could be trained on features with visibility-based data, it might achieve comparable reconstruction results with fewer views.

![Image](https://github.com/user-attachments/assets/d2826905-083a-40b5-8aa5-b0cd57f45937)

**Comment by vidit98** [comment]
Thanks a lot for sharing!

**Comment by DiamondGlassDrill** [comment]
Great thoughts and great expriment!

Does that mean @cyw-3d that we could as well train on material features along visibility features? Like in tripo3d, sf3d?

Cheers

**Comment by xiaoc57** [comment]
> I tried incorporating visibility into DINO feature aggregation using the PyTorch3D point cloud rasterizer to calculate visibility. I tested VAE reconstruction on 15 views, with visibility-based aggregation shown on the right of the figure and average aggregation on the left. The results appear quite similar. I suspect this is because the pretrained VAE I used was trained on average aggregation data, which contains multiple sources of noise. If a VAE could be trained on features with visibility-based data, it might achieve comparable reconstruction results with fewer views.
> 
> ![Image](https://github.com/user-attachments/assets/d2826905-083a-40b5-8aa5-b0cd57f45937)

Hi, could you please share the code for your visibility process? I have tried many methods but could not reproduce the effect you show.

**Comment by HaFred** [comment]
This should be a GS output mview frame clips.

**Comment by cyw-3d** [comment]
![Image](https://github.com/user-attachments/assets/bf2f84eb-925f-43fc-aa61-2623354a069c)
@xiaoc57 The visibility mask calculation code is shown in this figure.

**Comment by xiaoc57** [comment]
> ![Image](https://github.com/user-attachments/assets/bf2f84eb-925f-43fc-aa61-2623354a069c) [@xiaoc57](https://github.com/xiaoc57) The visibility mask calculation code is shown in this figure, it is added in the original extract_feature.py. Please let me know if you need the full python file and I will email you.

Hello, I do have some difficulties in reproducing. Please send the code to this email: 2415777318c@gmail.com. Thank you very much.

**Comment by Rody-NKCS** [comment]
@cyw-3d Hi,

Hope you're doing well.

I saw your very helpful note in the GitHub issue regarding the implementation of the visibility mask calculation in the extract_feature.py file.

Could you please kindly share the full extract_feature.py file containing this added visibility mask code? I would be extremely grateful to get a chance to look at the implementation in context.

My email address for sending the file is: ​​[Your Email Address Here]​​.

Thanks so much for considering this request, and for sharing your insights in the issue!

Best regards

**Comment by Rody1911641** [comment]
Hi @cyw-3d ,

Hope you're doing well.

I saw your very helpful note in the GitHub issue regarding the implementation of the visibility mask calculation in the extract_feature.py file.

Could you please kindly share the full extract_feature.py file containing this added visibility mask code? I would be extremely grateful to get a chance to look at the implementation in context.

My email address for sending the file is: ​​[Your Email Address Here]​​.

Thanks so much for considering this request, and for sharing your insights in the issue!

Best regards

**Comment by Luh1124** [comment]
> [@Rody1911641](https://github.com/Rody1911641) hello, I put together a code to visualize feature aggregation using fvdb. You can use it as a reference. I received help from [@cyw-3d](https://github.com/cyw-3d) in the implementation, and I would like to express my gratitude. The project is here [trellis_fvdb](https://github.com/xiaoc57/TRELLIS_fvdb/tree/master/visual_feature_aggregation). I used color to represent the features for visualization, you can easily modify it to dino features.

Hello, I tried changing the H/W resolution to 37 in your repo (Trellis_fvdb) to back-project DINOv2 features. After reconstructing with a VAE, significant floaters appeared. Would it be possible to share the source code with me by @cyw-3d? Or could you provide some guidance? My email is [645008691@qq.com](mailto:645008691@qq.com).

https://github.com/user-attachments/assets/89b271ab-0054-4261-ba49-00bfb1be94f4

**Comment by Luh1124** [comment]
> ![Image](https://github.com/user-attachments/assets/bf2f84eb-925f-43fc-aa61-2623354a069c) [@xiaoc57](https://github.com/xiaoc57) The visibility mask calculation code is shown in this figure, it is added in the original extract_feature.py. Please let me know if you need the full python file and I will email you.

Hello, I've tried multiple solutions, but I always get floaters. Could you also share a complete copy of the code with me?

**Comment by Luh1124** [comment]
@cyw-3d > ![Image](https://github.com/user-attachments/assets/bf2f84eb-925f-43fc-aa61-2623354a069c) [@xiaoc57](https://github.com/xiaoc57) The visibility mask calculation code is shown in this figure, it is added in the original extract_feature.py. Please let me know if you need the full python file and I will email you.

Hello, I've tried multiple solutions, but I always get floaters. Could you also share a complete copy of the code with me?

**Comment by Sugercanee** [comment]
> ![Image](https://github.com/user-attachments/assets/bf2f84eb-925f-43fc-aa61-2623354a069c) [@xiaoc57](https://github.com/xiaoc57) The visibility mask calculation code is shown in this figure, it is added in the original extract_feature.py. Please let me know if you need the full python file and I will email you.

Hi @cyw-3d ,

Hope you're doing well.

I saw your very helpful note in the GitHub issue regarding the implementation of the visibility mask calculation in the extract_feature.py file.

Could you please kindly share the full extract_feature.py file containing this added visibility mask code? I would be extremely grateful to get a chance to look at the implementation in context.

My email address for sending the file is: achen_7513@163.com.

Thanks so much for considering this request, and for sharing your insights in the issue!

Best regards

**Comment by xishuxishu** [comment]
> ![Image](https://github.com/user-attachments/assets/bf2f84eb-925f-43fc-aa61-2623354a069c) [@xiaoc57](https://github.com/xiaoc57) The visibility mask calculation code is shown in this figure, it is added in the original extract_feature.py. Please let me know if you need the full python file and I will email you.

Hi @cyw-3d , Thank you for your helpful explanation in this GitHub issue about the visibility mask calculation in extract_feature.py. I’m very interested in this part of the implementation, and I was wondering whether you might be willing to share the full extract_feature.py file that includes the visibility mask code. It would be really helpful for me to understand how this part is implemented in the full context of the pipeline.
 If it is possible to share, my email is: 2501212822 [at] stu.pku.edu.cn Thank you again for your time and for the helpful discussion here.

 Best regards

**Comment by vgoel-sc** [comment]
Hi, was anyone able to reproduce the results shown by @cyw-3d on VAE reconstruction with a less renders?

**Comment by xiaoc57** [comment]
> Hi, was anyone able to reproduce the results shown by [@cyw-3d](https://github.com/cyw-3d) on VAE reconstruction with a less renders?

Hello. Your problem seems to be about constructing voxel features using a small number of images, rather than considering image visibility. Constructing sparse voxel features based on visibility leads to a certain degree of quality degradation. Voxel features constructed using a small number of images result in out-of-distribution features that are difficult for VAE to process directly.
